goftests/utils.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2015-2016, Gamelan Labs, Inc.
+# Copyright (c) 2016, Google, Inc.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+# - Redistributions of source code must retain the above copyright
+#   notice, this list of conditions and the following disclaimer.
+# - Redistributions in binary form must reproduce the above copyright
+#   notice, this list of conditions and the following disclaimer in the
+#   documentation and/or other materials provided with the distribution.
+# - Neither the name of Salesforce.com nor the names of its contributors
+#   may be used to endorse or promote products derived from this
+#   software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+# TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+This module computes the chi^2 survival function and the required
+functions.
+
+This was implemented so that we could use the goftests module
+with pypy. Given that the only scipy dependence inside goftests
+was from the chi^2 survival function (1 - cdf), and then gamma
+function.
+
+"""
+
+from __future__ import division
+import math
+import sys
+from numpy import inf
+
+
+def log(x):
+    if x > sys.float_info.min:
+        value = math.log(x)
+    else:
+        value = -inf
+    return value
+
+
+def incomplete_gamma(x, s):
+    r"""
+    This function computes the incomplete lower gamma function
+    using the series expansion:
+    \gamma(x, s) = x^s \Gamma(s)e^{-x}\sum^\infty_{k=0}
+                    \frac{x^k}{\Gamma(s + k + 1)}
+    This series will converge strongly because the Gamma
+    function grows factorially.
+
+    Because the Gamma function does grow so quickly, we can
+    run into numerical stability issues. I have chosen to carry
+    out as much math as possible in the log domain to reduce
+    numerical error. This function matches the results from
+    scipy to numerical precision.
+    """
+    value = 0.0
+    if x < 0.0:
+        return 1.0
+    if x > 1e3:
+        return math.gamma(s)
+    log_gamma_s = log(math.gamma(s))
+    log_x = log(x)
+    for k in range(0, 100):
+        log_num = (k + s)*log_x + (-x) + log_gamma_s
```

Comment:
If the only use for `k` is in the expression `k + s`, you might want to just do `for k in range(s, s + 100):`